Add metadata to Razor

rynowak opened the pull request:
This PR introduces standard metadata to Razor. This change begins to
formalize the contract between generated code produced by Razor and
runtimes that want to load and interact with this code.

This is a step on making MVC a 'plugin' to Razor rather than the only
possible implementation. Since we're doing MSBuild work - this is the
right time to designate the current interaction between Razor and MVC as
'legacy' and move forward.

Additionally, we need the source thumbprinting to make re-compilation of
Razor a thing.

ajaybhargavb commented on src/Microsoft.AspNetCore.Razor.Runtime/Hosting/IRazorSourceThumprintMetadata.cs:
```diff
@@ -0,0 +1,16 @@
+
+namespace Microsoft.AspNetCore.Razor.Hosting
+{
+    public interface IRazorSourceThumprintMetadata
```
Why interface over abstract class?

rynowak commented:
> formalize the contract between generated code produced by Razor and runtimes that want to load and interact with this code.

I should mention that this is pretty similar to what we're doing today with the 'manifest'. Razor will generate a bunch of assembly attributes that identify which types are 'Razor' types. The `RazorCompiledItemAttribute` will include enough info to tell us what 'kind' of thing we're deal with and its 'identifier' - which is an analogue for the view engine path. 

This way we can easily scan and find, say, all of the pages, **OR** we can look up a specific one by path. From there we can fan out and look at metadata on the types themselves if we want to. The idea here is that at no point do we have to do an `Assembly.GetExportedTypes` 👍 

This system is also flexible enough for other non-MVC systems to use as long as the output is assemblies and types.

rynowak commented on src/Microsoft.AspNetCore.Razor.Runtime/Hosting/RazorSourceThumbprintAttribute.cs:
```diff
@@ -0,0 +1,74 @@
+    /// <see cref="RazorCompiledItemExtensions.GetThumprintMetadata(RazorCompiledItem)"/>.
+    /// </para>
+    /// </remarks>
+    [AttributeUsage(AttributeTargets.Class, AllowMultiple = true, Inherited = true)]
```
I was torn on whether or not to include other type-like things here other than class, so YAGNI kicked in. If someone wants to make an argument, I can be convinced.

rynowak replied:
Other things would be enum, interface, struct, etc.

ajaybhargavb commented on src/Microsoft.AspNetCore.Razor.Runtime/Hosting/RazorSourceThumbprintAttribute.cs:
```diff
@@ -0,0 +1,74 @@
+        /// <summary>
+        /// Gets the kind of the source file associated with this thumbprint.
+        /// </summary>
+        public string Kind { get; }
```
why not an enum for Kind? Same question for HashAlgorithm

rynowak replied:
Because we don't know all of the kinds that exist. MVC defines it's own kinds.

For hashalgorithm it might be possible to use an enum, but it's more coupling. If MVC wants to use a hash algorithm that Razor doesn't know then that requires a code change in Razor. That's bad coupling.

Think of it this way, it's your data, we're just holding on to it for you 😆

ajaybhargavb commented on src/Microsoft.AspNetCore.Razor.Runtime/Hosting/RazorCompiledItemLoader.cs:
```diff
@@ -0,0 +1,83 @@
+        /// </summary>
+        /// <param name="assembly">The <see cref="Assembly"/> to search.</param>
+        /// <returns>A list of <see cref="RazorCompiledItemAttribute"/> attributes.</returns>
+        protected IEnumerable<RazorCompiledItemAttribute> LoadAttributes(Assembly assembly)
```
protected virtual?

rynowak replied:
I didn't make this virtual because I don't think there's a different policy that works here. What's an example of a case where you'd override this?

ajaybhargavb commented on src/Microsoft.AspNetCore.Razor.Runtime/Hosting/RazorCompiledItem.cs:
```diff
@@ -0,0 +1,40 @@
+        /// For items loaded with the default implementation of <see cref="RazorCompiledItemLoader"/>, the 
+        /// metadata collection will return all attributes defined on the <see cref="Type"/>.
+        /// </remarks>
+        public IReadOnlyList<object> Metadata { get; }
```
abstract?

rynowak replied:
Yeah, good catch, this deffo should be abstract

rynowak commented on src/Microsoft.AspNetCore.Razor.Runtime/Hosting/RazorCompiledItemLoader.cs:
```diff
@@ -0,0 +1,83 @@
+    /// with additional data members by overriding <see cref="CreateItem(RazorCompiledItemAttribute)"/>.
+    /// </para>
+    /// </remarks>
+    public class RazorCompiledItemLoader
```
Also want to mention here, this is not an abstraction. This is infrastructure.

rynowak commented:
Updated and removed [Design]. This is implemented now end-to-end. Do yourself a favor and look at it commit-by-commit. There's lots of baseline updates, every runtime baseline gets updated as part of this change.

rynowak commented on src/Microsoft.AspNetCore.Razor.Language/DefaultRazorCodeGenerationOptionsFeature.cs:
```diff
@@ -22,15 +22,15 @@ protected override void OnInitialized()
         {
-            var builder = new DefaultRazorCodeGenerationOptionsBuilder(_designTime);
+            return _designTime ? RazorCodeGenerationOptions.CreateDesignTime(ConfigureOptions) : RazorCodeGenerationOptions.Create(ConfigureOptions);
+        }
```
This was calling the wrong API to create the code generation options

rynowak commented on src/Microsoft.AspNetCore.Razor.Language/RazorCodeDocumentExtensions.cs:
```diff
@@ -3,13 +3,61 @@
+        internal static readonly string IdentifierKey = "identifier";
+        internal static readonly string ImportIdentifiersKey = "imports-identifiers";
+
+        public static string GetIdentifier(this RazorCodeDocument document)
```
There's a big decision to make here about whether or not this should be part of the source document. After finishing this PR, I feel like it should be part of the source document. I'm planning to give that to @ajaybhargavb to play with. Thoughts?

rynowak replied:
The identifier is whatever you want. It's a way for the project and runtime to communicate. By default this will be a relative path.

rynowak commented on test/Microsoft.AspNetCore.Razor.Test.Common/Language/IntegrationTests/IntegrationTestBase.cs:
```diff
@@ -35,7 +35,7 @@ protected IntegrationTestBase()
         protected bool GenerateBaselines { get; set; } = true;
 #else
-        protected bool GenerateBaselines { get; set; } = false;
+        protected bool GenerateBaselines { get; set; } = true;
```
OOPS lol

rynowak commented on test/Microsoft.AspNetCore.Razor.Test.Common/Language/IntegrationTests/IntegrationTestBase.cs:
```diff
@@ -227,45 +232,5 @@ protected void AssertSourceMappingsMatchBaseline(RazorCodeDocument document)
             Assert.Equal(baseline, actual);
         }
-
-        protected class ApiSetsIRTestAdapter : IntermediateNodePassBase, IRazorOptimizationPass
```
Doing this in an optimization pass is the wrong thing. I reworked this feature to use the default document classifier, which is how we'd tell users to do it.

rynowak replied:
Also I had to do something here, because the `MetadataAttributePass` was seeing the uninitialized values.

rynowak commented on src/Microsoft.AspNetCore.Razor.Language/RazorCodeGenerationOptionsBuilder.cs:
```diff
@@ -11,8 +11,35 @@ public abstract class RazorCodeGenerationOptionsBuilder
+        /// a reference to <c>Microsoft.AspNetCore.Razor.Runtime</c>, or for testing purposes.
+        /// </para>
+        /// </remarks>
+        public virtual bool SuppressMetadataAttributes { get; set; }
```
Added docs here because the metadata attributes also include checksums. I wanted to make it clear which one does which thing

NTaylorMullen commented on src/Microsoft.AspNetCore.Razor.Language/Extensions/MetadataAttributePass.cs:
```diff
@@ -0,0 +1,112 @@
+            var insert = (int?)null;
+            for (var j = 0; j < @namespace.Children.Count; j++)
+            {
+                if (object.ReferenceEquals(@namespace.Children[j], @class))
```
Should we just add an intermediate node collection called `Attributes` to our `ClassDeclarationIntermediateNode`? Could even add it to the `MemberDeclrationIntermediateNode` if we want to enable it everywhere 😉

rynowak replied:
I started there but it's really a lot of work. Modeling attributes using a node has some challenges. You could do by adding the collection and not adding an Attribute node, but meh.

NTaylorMullen commented on src/Microsoft.AspNetCore.Razor.Language/RazorCodeDocumentExtensions.cs:
```diff
@@ -3,13 +3,61 @@
+            document.Items[IdentifierKey] = identifier;
+        }
+
+        public static IReadOnlyList<string> GetImportIdentifiers(this RazorCodeDocument document)
```
This goes away once @ajaybhargavb moves the `GetIdentifier` to source document yes?

NTaylorMullen commented on src/Microsoft.AspNetCore.Razor.Runtime/Hosting/RazorCompiledItemLoader.cs:
```diff
@@ -0,0 +1,83 @@
+            }
+
+            var items = new List<RazorCompiledItem>();
+            foreach (var attribute in LoadAttributes(assembly))
```
nit: split the `LoadAttributes` into its own local variable for easier debugging.

rynowak commented on src/Microsoft.AspNetCore.Razor.Language/CodeGeneration/DefaultDocumentWriter.cs:
```diff
@@ -67,28 +68,52 @@ public override void VisitDocument(DocumentIntermediateNode node)
-                    foreach (var value in checksum)
+                    string algorithmId;
+                    var algorithm = Context.SourceDocument.GetChecksumAlgorithm();
+                    if (algorithm == HashAlgorithmName.SHA256.Name)
```
Sure beans.

rynowak commented on src/Microsoft.AspNetCore.Razor.Language/CodeGeneration/DefaultDocumentWriter.cs:
```diff
@@ -67,28 +68,52 @@ public override void VisitDocument(DocumentIntermediateNode node)
+                            supportedAlgorithms,
+                            nameof(RazorCodeGenerationOptions) + "." + nameof(RazorCodeGenerationOptions.SuppressChecksum),
+                            bool.TrueString);
+                        throw new InvalidOperationException(message);
```
This is a programming error though. You've written a source document implementation that can't generate checksums we support.

rynowak replied:
Discussed in person, and I still think I'm right. Going to leave this as an exception.

pranavkm commented on src/Microsoft.AspNetCore.Razor.Language/Extensions/MetadataAttributeTargetExtension.cs:
```diff
@@ -0,0 +1,63 @@
+            context.CodeWriter.Write("[assembly: ");
+            context.CodeWriter.Write(CompiledItemAttributeName);
+            context.CodeWriter.Write("(typeof(");
+            context.CodeWriter.Write(node.TypeName);
```
Does this need `global::`?

rynowak replied:
Not here. We apply global to `node.TypeName` instead.

pranavkm commented on src/Microsoft.AspNetCore.Razor.Runtime/Hosting/RazorCompiledItem.cs:
```diff
@@ -0,0 +1,40 @@
+        /// For items loaded with the default implementation of <see cref="RazorCompiledItemLoader"/>, the 
+        /// metadata collection will return all attributes defined on the <see cref="Type"/>.
+        /// </remarks>
+        public abstract IReadOnlyList<object> Metadata { get; }
```
Neat

rynowak replied:
YEAHHHHHHHHHH BOI

rynowak commented on src/Microsoft.AspNetCore.Razor.Language/Extensions/IMetadataAttributeTargetExtension.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace Microsoft.AspNetCore.Razor.Language.Extensions
+{
+    internal interface IMetadataAttributeTargetExtension : ICodeTargetExtension
```
internal because there's no reason for it to be public. Just following the pattern we use elsewhere

rynowak commented on src/Microsoft.AspNetCore.Razor.Language/Extensions/MetadataAttributePass.cs:
```diff
@@ -0,0 +1,112 @@
+            var @namespace = documentNode.FindPrimaryNamespace();
+            if (@namespace == null || string.IsNullOrEmpty(@namespace.Content))
+            {
+                // No namespace node or it's incomplete. Skip.
```
Yeah, I'm still not convinced that's right. We need to do **something**, but I don't love that either.

rynowak commented on src/Microsoft.AspNetCore.Razor.Language/Extensions/MetadataAttributePass.cs:
```diff
@@ -0,0 +1,112 @@
+            // Now we need to add a [RazorSourceChecksum] for the source and for each import
+            // these are class attributes, so we need to find the insertion point to put them
+            // right before the class.
+            var insert = (int?)null;
```
that's probably true.

rynowak commented on test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/IncompleteDirectives_Runtime.codegen.cs:
```diff
@@ -1,6 +1,7 @@
 // <auto-generated/>
 #pragma warning disable 1591
+[assembly: global::Microsoft.AspNetCore.Razor.Hosting.RazorCompiledItemAttribute(typeof(AspNetCore.TestFiles_IntegrationTests_CodeGenerationIntegrationTest_IncompleteDirectives_cshtml), @"mvc.1.0.razor-page", @"TestFiles/IntegrationTests/CodeGenerationIntegrationTest/IncompleteDirectives")]
 [assembly:global::Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.RazorPageAttribute(null, typeof(AspNetCore.TestFiles_IntegrationTests_CodeGenerationIntegrationTest_IncompleteDirectives_cshtml), null)]
```
Yes. This is the new contract between Razor and MVC

rynowak commented on src/Microsoft.AspNetCore.Razor.Language/Extensions/MetadataAttributePass.cs:
```diff
@@ -0,0 +1,112 @@
+        
+        // Optimization pass is the best choice for this class. It's not an optimization, but it also doesn't add semantically
+        // meaningful information.
+        IRazorOptimizationPass
```
Sure

rynowak commented:
Going to merge this and then take a look at that travis failure, which is not related to this PR.
